### PR TITLE
[consensus] Make CommitRange inclusive and use in commit syncer

### PR DESCRIPTION
**Pull request by arun-koshy**
## Test plan 

pending ptn run

### CODE REVIEW

File: consensus/core/src/network/tonic_network.rs [hunk 1]
```diff
@@ -224,12 +225,14 @@ impl NetworkClient for TonicClient {
...
         &self,
         peer: AuthorityIndex,
-        start: CommitIndex,
-        end: CommitIndex,
+        commit_range: CommitRange,
         timeout: Duration,
     ) -> ConsensusResult<(Vec<Bytes>, Vec<Bytes>)> {
         let mut client = self.get_client(peer, timeout).await?;
-        let mut request = Request::new(FetchCommitsRequest { start, end });
+        let mut request = Request::new(FetchCommitsRequest {
```
Comment (arun-koshy) [inline consensus/core/src/network/tonic_network.rs:232]: note: decided against using `CommitRange` in `FetchCommitsRequest` so we could keep the types of the message as `u32` which should be more efficient.

Comment (mwtian) [reply consensus/core/src/network/tonic_network.rs:232]: Yes that would be a backward incompatible change.

Comment (arun-koshy) [reply consensus/core/src/network/tonic_network.rs:232]: oh right! That should have been the main reason not to do it 😆

File: consensus/core/src/authority_service.rs [hunk 2]
```diff
@@ -314,15 +314,17 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
...
     ) -> ConsensusResult<(Vec<TrustedCommit>, Vec<VerifiedBlock>)> {
         fail_point_async!("consensus-rpc-response");
 
-        // Compute an exclusive end index and bound the maximum number of commits scanned.
-        let exclusive_end =
-            (end + 1).min(start + self.context.parameters.commit_sync_batch_size as CommitIndex);
-        let mut commits = self.store.scan_commits((start..exclusive_end).into())?;
+        // Compute an inclusive end index and bound the maximum number of commits scanned.
+        let inclusive_end = (commit_range.end()).min(
+            commit_range.start() + self.context.parameters.commit_sync_batch_size as CommitIndex,
```
Comment (mwtian) [inline consensus/core/src/authority_service.rs:323]:
```suggestion
            commit_range.start() + self.context.parameters.commit_sync_batch_size as CommitIndex - 1,
```

File: consensus/core/src/commit.rs [hunk 3]
```diff
@@ -543,30 +543,30 @@ impl CommitInfo {
...
     pub(crate) fn end(&self) -> CommitIndex {
-        self.0.end
+        *self.0.end()
     }
 
     /// Check if the provided range is sequentially after this range with the same
     /// range length.
     pub(crate) fn is_next_range(&self, other: &Self) -> bool {
-        self.0.len() == other.0.len() && self.end() == other.start()
+        self.0.size_hint() == other.0.size_hint() && self.end() + 1 == other.start()
```
Comment (mwtian) [inline consensus/core/src/commit.rs:569]: Why did we need to check size again? Seems a bit unexpected?

File: consensus/core/src/commit.rs [hunk 4]
```diff
@@ -543,30 +543,30 @@ impl CommitInfo {
...
 
 /// CommitRange stores a range of CommitIndex. The range contains the start (inclusive)
-/// and end (exclusive) commit indices and can be ordered for use as the key of a table.
-#[derive(Clone, Default, Debug, PartialEq, Eq, Serialize, Deserialize)]
-pub(crate) struct CommitRange(Range<CommitIndex>);
+/// and end (inclusive) commit indices and can be ordered for use as the key of a table.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct CommitRange(RangeInclusive<CommitIndex>);
 
 #[allow(unused)]
```
Comment (mwtian) [inline consensus/core/src/commit.rs:550]: Remove unused annotation?

File: consensus/core/src/commit_syncer.rs [hunk 5]
```diff
@@ -143,14 +144,14 @@ impl<C: NetworkClient> CommitSyncer<C> {
                         // Create range with inclusive start and end.
                         let range_start = prev_end + 1;
                         let range_end = prev_end + inner.context.parameters.commit_sync_batch_size;
-                        // When the condition below is true, [range_start, range_end] contains less number of commits
+                        // When the condition below is true, CommitRange(range_start..=range_end) contains less number of commits
```
Comment (mwtian) [inline consensus/core/src/commit_syncer.rs:147]: Revert? The original representation conveys that the start and end are inclusive.

File: consensus/core/src/commit_syncer.rs [hunk 6]
```diff
@@ -347,10 +342,14 @@ impl<C: NetworkClient> CommitSyncer<C> {
             sleep(available_time - now).await;
         }
 
-        // 2. Fetch commits in the range [start, end] from the selected authority.
+        // 2. Fetch commits in CommitRange(start..=end) from the selected authority.
```
Comment (mwtian) [inline consensus/core/src/commit_syncer.rs:345]:
```suggestion
        // 2. Fetch commits in the commit range from the selected authority.
```

File: consensus/core/src/network/mod.rs [hunk 7]
```diff
@@ -94,14 +94,13 @@ pub(crate) trait NetworkClient: Send + Sync + Sized + 'static {
         timeout: Duration,
     ) -> ConsensusResult<Vec<Bytes>>;
 
-    /// Fetches serialized commits from a peer, with index in [start, end].
+    /// Fetches serialized commits from a peer, with index in CommitRange(start..=end).
```
Comment (mwtian) [inline consensus/core/src/network/mod.rs:97]:
```suggestion
    /// Fetches serialized commits in the commit range from a peer.
```

File: consensus/core/src/commit.rs (hunk 3 above)
Comment (arun-koshy) [reply consensus/core/src/commit.rs:572]: We want to ensure that the next range starts after this range and is of the same size. This is an assertion we want to check before we update the leader swap table. RangeInclusive does not have len() method so we had to use size_hint which checks the steps between start and end.

Comment (mwtian) [reply consensus/core/src/commit.rs:571]: It does not seem very intuitive that `is_next_range()` also expects the sizes are the same. What about just making sure `self.end() + 1 == other.start()` here, and do the size check at the callsite? Or move this function to leader schedule.

File: consensus/core/src/commit.rs [hunk 8]
```diff
@@ -546,30 +546,29 @@ impl CommitInfo {
 }
 
 /// CommitRange stores a range of CommitIndex. The range contains the start (inclusive)
-/// and end (exclusive) commit indices and can be ordered for use as the key of a table.
-#[derive(Clone, Default, Debug, PartialEq, Eq, Serialize, Deserialize)]
-pub(crate) struct CommitRange(Range<CommitIndex>);
+/// and end (inclusive) commit indices and can be ordered for use as the key of a table.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
```
Comment (mwtian) [inline consensus/core/src/commit.rs:550]: `Copy` can be derived as well, so we don't need to explicitly `clone()` CommitRange.

File: consensus/core/src/authority_service.rs [hunk 9]
```diff
@@ -314,15 +314,18 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
...
+        commit_range: CommitRange,
     ) -> ConsensusResult<(Vec<TrustedCommit>, Vec<VerifiedBlock>)> {
         fail_point_async!("consensus-rpc-response");
 
-        // Compute an exclusive end index and bound the maximum number of commits scanned.
-        let exclusive_end =
-            (end + 1).min(start + self.context.parameters.commit_sync_batch_size as CommitIndex);
-        let mut commits = self.store.scan_commits((start..exclusive_end).into())?;
+        // Compute an inclusive end index and bound the maximum number of commits scanned.
+        let inclusive_end = (commit_range.end()).min(
```
Comment (mwtian) [inline consensus/core/src/authority_service.rs:322]:
```suggestion
        let inclusive_end = commit_range.end().min(
```

File: consensus/core/src/commit.rs (hunk 8 above)
Comment (arun-koshy) [reply consensus/core/src/commit.rs:550]: Unfortunately `RangeInclusive` does not impl `Copy`

File: consensus/core/src/commit.rs [hunk 10]
```diff
@@ -694,11 +699,19 @@ mod tests {
         assert_eq!(range1.start(), 1);
         assert_eq!(range1.end(), 5);
 
+        tracing::debug!("test");
```
Comment (mwtian) [inline consensus/core/src/commit.rs:702]: Remove?

**Review by mwtian — APPROVED** [review]

**Pull request merged**